ttnn/cpp/ttnn/operations/ccl/barrier/device/barrier_op.cpp

```diff
@@ -0,0 +1,98 @@
+    return std::vector<SimpleShape>(input_tensors.size(), shape);
+}
+
+std::vector<Tensor> Barrier::create_output_tensors(const std::vector<Tensor>& input_tensors) const {
+    //Tensor is unmodified, return what was passed in
+    return input_tensors;
+}
+
+operation::ProgramWithCallbacks Barrier::create_program(
+    const std::vector<Tensor>& input_tensors, std::vector<Tensor>& output_tensors) const {
+    return ccl::barrier::detail::barrier_with_workers(
+        input_tensors.at(0),
+        output_tensors.at(0),
+        this->is_starting_core,
+        this->ring_size,
+        this->ring_index,
+        this->receiver_device_id,
+        this->sender_device_id,
+        this->topology);
+}
+
+void Barrier::update_structure (const Tensor& input_tensor)
+{
+    //Need to resolve the neighbours of this tensor
+    //Can only be done in launch_op as it differs for different input tensors
+    const auto devices = this -> devices;
+    const bool is_linear = (topology == ttnn::ccl::Topology::Linear);
+    const uint32_t num_devices = this->ring_size;
+    uint32_t device_index = 0;
+    std::optional<chip_id_t> receiver_device_id = std::nullopt;
+    std::optional<chip_id_t> sender_device_id = std::nullopt;
+    for (uint32_t i = 0; i < num_devices; ++i) {
+        if (devices.at(i) == input_tensor.device()) {
+            bool is_last_chip_in_clockwise_direction = is_linear && i == (num_devices - 1);
+            bool is_last_chip_in_counter_clockwise_direction = is_linear && i == 0;
+            device_index = i;
+            receiver_device_id = is_last_chip_in_clockwise_direction ?
+                std::nullopt :
+                std::optional<chip_id_t>(devices.at((i + 1) % num_devices)->id());
+            sender_device_id = is_last_chip_in_counter_clockwise_direction ?
+                std::nullopt :
+                std::optional<chip_id_t>(devices.at((i + num_devices - 1) % num_devices)->id());
+            break;
+        }
+    }
+    this ->receiver_device_id = receiver_device_id;
+    this ->sender_device_id = sender_device_id;
+    this -> ring_index = device_index;
+    this -> is_starting_core = device_index==0;
+}
+namespace operations::ccl{
+
+
+Tensor barrier(
+    const Tensor& input_tensor,
+    const ttnn::Barrier& barrier_struct)
+{
+    std::vector<Tensor> output_tensors = {Tensor(operation::get_workers_for_op_output({input_tensor}))};
+    operation::launch_op
+    (
+        [barrier_struct](
+            const std::vector<Tensor>& input_tensors,
+            const std::vector<std::optional<const Tensor>>& optional_input_tensors,
+            const std::vector<std::optional<Tensor>>& optional_output_tensors) mutable -> std::vector<Tensor>
+            {
+                const Tensor& input_tensor = input_tensors.at(0);
+                // need to copy and update barrier struct for this particular tensor
+                ttnn::Barrier new_barrier_struct = barrier_struct;
```

Comment:
Can you please explain this sequence? I'm a bit confused by it. 

It looks like we call `barrier(inpout_tensor, Barrier(..., input_tensor.get_workers())` then we have to manually update the barrier struct. Two things:

1) `barrier_struct` is captured by value/copy here so we shouldn't need this explicit copy
2) Why not just initialize the structure like this `update_structure` call but just do it in the constructor?

Reply:
I think my comment may be misleading. If I bring into the invoke function just a variable rather than a constant, I will see corruption between the different devices (the values seem to get overwritten for each subsequent device so in runtime I have the wrong config being deployed). I suspect that internally there must be some form of shallow copying and every time this function is called for each of the devices, we may be overwritting the same memory structure which can lead to a race condition if the device has not loaded in that data when the compiler runs launch_op for a different device. Since it is a constant, I need to be internally creating a copy of the structure that isn't constant so I can update it with the device specific stuff.

Reply:
Now I guess I could improve compilation time by a few nanoseconds by instead creating a different barrier_parameters structure that only contains the non device-specific parameters, passing that in, and then combining update_structure with barrier structure initialization which would take in this barrier_parameters structure as an input. However I personally think the extra copying time of doing it this way rather than copying in the leaner barrier_parameters does not outweigh the cost of making this flow harder to follow. And since it is on the compile and not runtime side I didn't think a few nanoseconds were significant, but if you disagree I can do that, it would only take a couple minutes.

Reply:
Thanks I see that the passed arg is const ref. okay sounds good. Don't bother splitting - I was just asking for clarification here.